ydb/src/credentials.rs

```diff
@@ -12,19 +12,116 @@ use http::Uri;
 use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
 use secrecy::{ExposeSecret, SecretString};
 use serde::{Deserialize, Serialize};
+use std::env;
 use std::fmt::Debug;
 use std::ops::Add;
 use std::process::Command;
 use std::sync::{Arc, Mutex};
 use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
-use tracing::{trace,debug};
+use tracing::{debug, trace};
+
+const YDB_ANONYMOUS_CREDENTIALS: &str = "YDB_ANONYMOUS_CREDENTIALS";
+const YDB_SERVICE_ACCOUNT_KEY_FILE_CREDENTIALS: &str = "YDB_SERVICE_ACCOUNT_KEY_FILE_CREDENTIALS";
+const YDB_METADATA_CREDENTIALS: &str = "YDB_METADATA_CREDENTIALS";
+const YDB_ACCESS_TOKEN_CREDENTIALS: &str = "YDB_ACCESS_TOKEN_CREDENTIALS";
+
+pub const YDB_METADATA_URL: &str =
+    "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token";
+
+const EMPTY_TOKEN: &str = "";
+
+pub type MetadataUrlCredentials = GCEMetadata;
+pub type AccessTokenCredentials = StaticToken;
+pub type CommandLineCredentials = CommandLineYcToken;
+pub type StaticCredentials = StaticCredentialsAuth;
 
 pub(crate) type CredentialsRef = Arc<Box<dyn Credentials>>;
 
 pub(crate) fn credencials_ref<T: 'static + Credentials>(cred: T) -> CredentialsRef {
     Arc::new(Box::new(cred))
 }
 
+pub struct AnonymousCredentials {
+    inner: StaticToken,
+}
+
+impl AnonymousCredentials {
+    pub fn new() -> Self {
+        Self {
+            inner: StaticToken::from(EMPTY_TOKEN),
+        }
+    }
+}
+
+impl Credentials for AnonymousCredentials {
+    fn create_token(&self) -> YdbResult<TokenInfo> {
+        self.inner.create_token()
+    }
+}
+
+/// Select credentials from environment
+/// reference: https://ydb.tech/docs/en/reference/ydb-sdk/auth
+pub struct FromEnvCredentials {
+    inner: InnerCredentials,
```

Comment:
What about use trait Credentials instead of enum?

Or make FromEnvCredentials as function without additional struct?

pub fn new_credentials_from_env()->YdbResult<Box<dyn Credentials>>

Reply:
OK, I'll rewrite it